File: rottingOranges.py

# Time Complexity : O(n) - n = size of grid
# Space Complexity : O(n) - n= size of grid
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        fresh = 0
        if grid is None:
            return 0
        q = deque()
        m,n = len(grid),len(grid[0])
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh+=1
                elif grid[i][j] == 2:
                    q.append((i,j))
        if fresh == 0:
            return 0;
        dirs = [(0,1),(0,-1),(1,0),(-1,0)]
        time = 0
        while q:
            size = len(q)
            for j in range(size):
                curr = q.popleft()
                for i in dirs:
                    r,c = i[0]+curr[0],i[1]+curr[1]
                    if m>r>=0 and n>c>= 0:
                        if grid[r][c] == 1:
                            grid[r][c]=2
                            fresh-=1
                            q.append((r,c))
            time+=1
        return time-1 if fresh == 0 else -1
            

File: RottingOranges.cs
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

//BFS Approach
	// 1) BFS
	// 2) Put al rotten oranges in queue
	// 3) Pop the prange
	// 4) Look for neighbors, 
	// 	a. If they are fresh make them rotten
	// 	b. Put in queue
	// 	c. Kepp repeating the loop until queue is empty

//O(n), O(n)
 int m, n;
int[][] dirs;

public int OrangesRotting(int[][] grid) {
    if(grid == null || grid.Length == 0)
        return 0;
    
    int time = 0;
    int fresh = 0;
    m = grid.Length;
    n = grid[0].Length;
    
    dirs = new int[][]{new int []{0, 1}, new int[]{1,0}, new int[]{0, -1}, new int[] {-1, 0}};
    
    Queue<int[]> queue = new Queue<int[]>();
    for(int i = 0; i < grid.Length; i++)
    {
        for(int j = 0; j < n; j++)
        {
            if(grid[i][j] == 2)
                queue.Enqueue(new int[] {i, j});
                
            if(grid[i][j] == 1)
                fresh++;
        }
    }
    
    if(fresh == 0)
        return 0;   
    
    while(queue.Count > 0)
    {
        int size = queue.Count;
        for(int i = 0; i < size; i++)
        {
            var curr = queue.Dequeue();
            foreach(var dir in dirs)
            {
                int nr = curr[0] + dir[0];
                int nc = curr[1] + dir[1];
                
                if(nr >= 0 && nc >= 0 && nr < m && nc < n && grid[nr][nc] == 1)
                {
                    grid[nr][nc] = 2;
                    queue.Enqueue(new int[]{nr, nc});
                    fresh--;
                }
            }
        }
        time++;
    }
        
    if(fresh != 0)
        return -1;
    //doing time -1 because we are doing time++, it is possible, there is no adjacent negibor to rotten it
    //but it will increase the time + 1, evev though no operation happened
    return time - 1;
}